0012-unify-pulse-commands-and-instructions.md

```diff
@@ -0,0 +1,334 @@
+|----------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `Instruction`  | This is not a user-facing class from an API interface perspective. This will be rewritten as per the outline in the detailed design section. The `name` parameter will be deprecated.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
+| `PulseCommand` | Will be converted to a `Pulse` and all pulse classes will inherit from this. Calling a `Pulse` with a `Channel` as input will now yield an instruction of the form `Play(pulse, channel)` this will enable the same syntax for instructions to be kept.                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
+| `Delay`        | Will be converted from a `Command` to a new `Instruction` type. For the deprecation period the `Channel` argument will be made optional but if not supplied will emit a warning and a `None` type will be set for the second operands attribute. Calling the instruction with a `Channel` will yield a new instruction with the parameters of the original instruction and the additional input `channel` operands being the input channel. If the instruction with a `channel` attribute of `None` makes it to assembly, an error shall be raised.                                                                                                                                                                               |
```

Comment:
I just want to be clear about how the `Delay` behaves. (Maybe this comment should not go for this RFC but for terra directly since `Delay` has already been provided, though.)

`Delay` and unoccupied timeslot is completely different in that `Delay` occupies a timeslot and cannot overlap with any other `Instruction`, right? For example,
```
p_len10 = Pulse(duration=10)

sched = Schedule()
sched = sched.append(Play(p_len10, DriveChannel(0)))
sched = sched.append(Play(p_len10, DriveChannel(0)) << 10)  # shift(10)
sched = sched.insert(10, Play(p_len10, DriveChannel(0)))  # Succeed!

# `shift`+`append` cannot be replaced with `Delay` (i.e. `Delay` occupies a timeslot)
sched = Schedule()
sched = sched.append(Play(p_len10, DriveChannel(0)))
sched = sched.append(Delay(10, DriveChannel(0)))
sched = sched.append(Play(p_len10, DriveChannel(0)))
sched = sched.insert(10, Play(p_len10, DriveChannel(0)))  # Fail!

# `rescheduler.pad()` usually implies no more `insert`
sched = Schedule()
sched = sched.append(Play(p_len10, DriveChannel(0)))
sched = sched.append(Play(p_len10, DriveChannel(0)) << 10)  # shift(10)
sched = rescheduler.pad(sched)
sched = sched.insert(10, Play(p_len10, DriveChannel(0)))  # Fail!
```

It would be helpful to add docstring (or examples in notebook) notifying the difference between Delay and {unoccupied timeslot, shift}.

Reply:
These docstrings do exist already, and regardless Timeslots will be disappearing soon in #3749.

Reply:
Oh, I see. I should have checked current `DelayInstruction` docstrings, it says
```
        No other command may be scheduled within a delay command.
```